File: clients/apps/lib/services/user_service.dart
import 'package:dio/dio.dart';
import '../models/models.dart';
import '../models/requests.dart';
import '../models/responses.dart';
import '../network/user_api_client.dart';

class UserService {
  final UserApiClient _client;
  final Dio _dio;

  UserService(this._client, this._dio);

  Future<User> getProfile() async {
    final res = await _client.getProfile();
    return res.data.data!;
  }

  Future<UserPublicProfile> getUserInfo(String userId) async {
    final res = await _client.getUserInfo(userId);
    return res.data.data!;
  }

  Future<List<UserSearchResult>> searchUsers(String query) async {
    final res = await _client.searchUsers(query);
    return res.data.data?.users ?? [];
  }

  Future<User> updateProfile(UpdateProfileRequest request) async {
    final res = await _client.updateProfile(request);
    return res.data.data!;
  }

  Future<UploadAvatarResponse> uploadAvatar(String filePath) async {
    final fileName = filePath.split(RegExp(r'[/\\]')).last;
    final formData = FormData.fromMap({
      'file': await MultipartFile.fromFile(filePath, filename: fileName),
    });
    final res = await _dio.post<Map<String, dynamic>>(
      '/user/upload',
      data: formData,
    );
    final body = res.data!;
    final data = body['data'] as Map<String, dynamic>;
    return UploadAvatarResponse.fromJson(data);
  }
}

File: clients/packages/shared/src/services/user.service.ts
import type {
  ApiResponse,
  SearchUsersResponse,
  UpdateProfileDTO,
  UploadImageResponse,
  User,
  UserPresence,
  UserPublicProfile,
} from '../types'
import { http } from './http'

export const userService = {
  async getProfile(): Promise<ApiResponse<User>> {
    const res = await http.get<ApiResponse<User>>('/user/me')
    return res.data
  },

  async updateProfile(data: UpdateProfileDTO): Promise<ApiResponse<User>> {
    const res = await http.put<ApiResponse<User>>('/user/me', data)
    return res.data
  },

  async searchUsers(
    query: string,
    limit = 20,
    signal?: AbortSignal,
  ): Promise<ApiResponse<SearchUsersResponse>> {
    const res = await http.get<ApiResponse<SearchUsersResponse>>('/user/search', {
      params: { q: query, limit },
      signal,
    })
    return res.data
  },

  async getUserInfo(userId: string): Promise<ApiResponse<UserPublicProfile>> {
    const res = await http.get<ApiResponse<UserPublicProfile>>(`/user/${userId}`)
    return res.data
  },

  async getPresenceBatch(userIds: string[]): Promise<ApiResponse<{ users: UserPresence[] }>> {
    const res = await http.post<ApiResponse<{ users: UserPresence[] }>>('/user/presence', { userIds })
    return res.data
  },

  async uploadImage(file: File): Promise<ApiResponse<UploadImageResponse>> {
    const formData = new FormData()
    formData.append('file', file)
    const res = await http.post<ApiResponse<UploadImageResponse>>(
      '/user/upload',
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } },
    )
    return res.data
  },
}
